Created "recursion" for COP 3502C

JacobGa03 opened the pull request:
Created some tutorials for the basics of recursion. I created a folder and added the .md files to the "cs1" folder. I also added the tutorials to the index.md too.

Thanks.

JacobGa03 committed:
Created "recursion" for COP 3502C

Created some tutorials for the basics of recursion.

JacobGa03 committed:
Merge branch 'main' into recursion-basics

JacobGa03 committed:
Merge branch 'main' into recursion-basics

jerrettl commented on src/cs1/recursion/recursion-basics.md:
```diff
@@ -0,0 +1,135 @@
+
+Let's consider a classic example that is used to teach recursion: Factorial!
+
+The factorial of any non-negative integer, `n`, usually written as `n!`, is defined as the product of all the positive-integers less than or equal to `n`.
+
+`n! = n * (n-1) * (n-2) * ... * 3 * 2 * 1`
```
Since you are working with a math equation here, maybe you want to use math notation format for this? Not vitally important, just a touch to consider. Also I may add a colon at the end of the line to connect the equation and sentence together:

```suggestion
The factorial of any non-negative integer, $n$, usually written as $n!$, is defined as the product of all the positive-integers less than or equal to $n$:

$n! = n * (n-1) * (n-2) * ... * 3 * 2 * 1$
```

JacobGa03 replied:
I've just plugged this in to play around. I really like the LaTex feel. I never considered doing this, but I think I will for the math explanations; it seems to take up less space and looks nicer (imo).

jerrettl commented on src/cs1/recursion/recursion-basics.md:
```diff
@@ -0,0 +1,135 @@
+
+This issue highlights another key pillar of recursive solutions: base cases. A base case is a case in our recursive solution that is trivially easy to solve. Because of this, we can simply return the answer to problem, this also ensures that our recursive calls end. For Factorial this base case will be `fact(0)`, since `0! = 1`.
+
+Now, lets add a base case to our code and see if it works
```
(This would be something a linter would say)

```suggestion
Now, let's add a base case to our code and see if it works:
```

jerrettl commented on src/cs1/recursion/recursion-basics.md:
````diff
@@ -0,0 +1,135 @@
+
+Let's consider writing an iterative solution to this problem. This would the solution someone who hasn't encountered recursion before would probably come up with:
+
+``` c
````
The first indentation level in code blocks should match where the tick marks are. (This comment would really be in a linter though.)

JacobGa03 replied:
Perhaps this is a VSCode setting that is messing this up? 

Are you saying something like this?

```c
   Here is some text
```

JacobGa03 replied:
I think this might be a VSCode settings for using tabs instead of spaces, but I'm not sure yet.

jerrettl commented on src/cs1/recursion/recursion-basics.md:
```diff
@@ -0,0 +1,135 @@
+
+## Formulating a Recursive Solution
+
+When creating a recursive solution we need to consider how we can create the smaller sub-problems. When computing the factorial, notice at each iteration we are simply multiplying the number we are at ,`n`, with whatever factorial of `n-1` is. 
```
Small English thing:

```suggestion
When creating a recursive solution we need to consider how we can create the smaller sub-problems. When computing the factorial, notice at each iteration we are simply multiplying the number we are at, `n`, with whatever factorial of `n-1` is. 
```

jerrettl commented on src/cs1/recursion/recursion-basics.md:
```diff
@@ -0,0 +1,135 @@
+
+When creating a recursive solution we need to consider how we can create the smaller sub-problems. When computing the factorial, notice at each iteration we are simply multiplying the number we are at ,`n`, with whatever factorial of `n-1` is. 
+
+Lets write a function body along with the logic I just described
```
Just a small suggestion for English reading on this line:

```suggestion
Let's write a function body using the logic I just described:
```

jerrettl commented on src/cs1/recursion/recursion-basics.md:
```diff
@@ -0,0 +1,135 @@
+
+`n! = n * (n-1) * (n-2) * ... * 3 * 2 * 1`
+
+Let's consider writing an iterative solution to this problem. This would the solution someone who hasn't encountered recursion before would probably come up with:
```
Maybe add a small explanation of "iterative solution" here? Not sure if necessary, and it doesn't have to be a very long explanation, but since this is introducing recursion, this may be good to show the contrast between them.

jerrettl replied:
This sentence was a little funny when I read it, there are multiple ways to do it but this is one way I thought up for phrasing:

```suggestion
Let's consider writing an iterative solution to this problem. This would probably be the solution from someone who hasn't encountered recursion before:
```

jerrettl commented on src/cs1/recursion/recursion-basics.md:
````diff
@@ -0,0 +1,135 @@
+
+Let's consider writing an iterative solution to this problem. This would the solution someone who hasn't encountered recursion before would probably come up with:
+
+``` c
````
Note here that `ans` is not initialized, so technically this could give an unexpected answer. (Assuming it was 1, this does work, but if it was 0 it would not.)

JacobGa03 replied:
Ahh good catch. I will make some changes to ensure there is no ambiguity.

jerrettl commented on src/cs1/recursion/recursion-basics.md:
```diff
@@ -0,0 +1,135 @@
+
+The key to utilizing recursive solutions to problems is answering one key question: "Can the problem I'm solving be broken into smaller sub-problems of the same type?"
+
+Notice for our simple Factorial program, we only do simple multiplication for each iteration `ans *= i`.
```
Maybe add some punctuation here for English prose's sake:

```suggestion
Notice for our simple Factorial program, we only do simple multiplication for each iteration: `ans *= i`.
```

jerrettl commented on src/cs1/recursion/recursion-basics-answers.md:
The file name here implies the questions are in `recursion-basics.md`, but this is stand-alone. This is part of the main "contributor guidelines" comments, but should problems and solutions be separated or combined? I am open to suggestion either way.

jerrettl commented on src/cs1/recursion/recursion-basics.md:
```diff
@@ -0,0 +1,135 @@
+
+Well let's trace through `fact(5)`
+
+> When things don't work like intended, it is always helpful to run a debugger, like GDB, which is extremely useful to show the steps your code took during execution. WikiKnights has tutorials on using GDB if you are interested.
```
When I was reading through, the block felt a little like it was in the middle of a thought with tracing. I may suggest moving it after the thought is finished. Or you could also add a label like:

```suggestion
> **Note:** When things don't work like intended, it is always helpful to run a debugger, like GDB, which is extremely useful to show the steps your code took during execution. Wiki Knights has tutorials on using GDB if you are interested.
```

jerrettl commented on src/cs1/recursion/recursion-basics.md:
````diff
@@ -0,0 +1,135 @@
+                            .
+                            .
+                            .
+```
````
For fun, I would illustrate the fact that it goes to negatives here:

`````suggestion
```
                    fact(5) = 5 * fact(4)
                    fact(4) = 4 * fact(3)
                    fact(3) = 3 * fact(2)
                    fact(2) = 2 * fact(1)
                    fact(1) = 1 * fact(0)
                    fact(0) = 0 * fact(-1)
                    fact(-1) = -1 * fact(-2)
                            .
                            .
                            .
```
`````

jerrettl commented on src/cs1/recursion/recursion-basics.md:
````diff
@@ -0,0 +1,135 @@
+                            .
+```
+
+This issue highlights another key pillar of recursive solutions: base cases. A base case is a case in our recursive solution that is trivially easy to solve. Because of this, we can simply return the answer to problem, this also ensures that our recursive calls end. For Factorial this base case will be `fact(0)`, since `0! = 1`.
````
I may just add a note that 0! = 1 because it is defined:

```suggestion
This issue highlights another key pillar of recursive solutions: base cases. A base case is a case in our recursive solution that is trivially easy to solve. Because of this, we can simply return the answer to problem, this also ensures that our recursive calls end. For Factorial this base case will be `fact(0)`, since `0! = 1`, by definition.
```

jerrettl commented on src/cs1/recursion/recursion-basics.md:
````diff
@@ -0,0 +1,135 @@
+    }
+}
+```
+
````
What happens if you call `fact()` with a negative number? Some will argue the condition should be `if (n <= 0)` for this as a "just-in-case," even if it should never happen. Not necessary, but you could put a note here for the curious student.

jerrettl commented on src/cs1/recursion/recursion-basics.md:
````diff
@@ -0,0 +1,135 @@
+
+If we analyze the structure of our recursive solution to the factorial, we will notice a few things:
+
+```c
+ //Base case
+ if(terminating condition){
+    //Return the trivially easy answer
+ }
+ //Recursive case
+ else{
+    //Slightly solve and recursively call the function again
+ }
+
+```
````
Interesting that Dr. Guha uses this structure. The other common way is:

``` c
// Base case
if (terminating condition)
{
    // Return trivially easy answer
}

// Slightly solve and recursively call the function again
```

Since the if statement is returning, everything after will only ever execute if the condition was evaluated false. Maybe throw it in for reference and exposure?

jerrettl commented on src/cs1/recursion/recursion-basics.md:
````diff
@@ -0,0 +1,135 @@
+int fact(int n){
+    return n * factorial(n-1);
+}
+```
````
Just a small mismatch between `fact()` and `factorial()` here, either is good but should just be the same throughout the page. (This is not the only place)

jerrettl commented on src/cs1/recursion/recursion-basics.md:
````diff
@@ -0,0 +1,135 @@
+int sumDigits(int n){
+    // You write this
+}
+```
````
This is completely a technicality, but the function header is just the first line without the brackets. (I only mention this since this is a learning resource for students and I would prefer the terms to be rock-solid.) Many ways to make this consistent, this is just one:

```suggestion
Given the following function, write a recursive solution to this problem.

```c
int sumDigits(int n){
    // You write this
}
```
```

jerrettl commented on src/cs1/recursion/recursion-basics-answers.md:
````diff
@@ -0,0 +1,70 @@
+}
+
+```
+Here we calculate the `n'th` Lucas number.
````
Small thing, but `n'th` is inconsistent with `nth` earlier, this should just be consistent either way.

jerrettl commented on src/cs1/recursion/recursion-basics-answers.md:
````diff
@@ -0,0 +1,70 @@
+```
+Here we calculate the `n'th` Lucas number.
+
+Unlike our previous example, we have two different base cases. The reason for this comes from the fact that there are two initial Lucas numbers: `0` and `1`. To handle this, we write to se
````
I think this may not be fully pushed to the branch yet? This ends mid-sentence 😅

jerrettl commented on src/cs1/recursion/recursion-basics.md:
```diff
@@ -0,0 +1,135 @@
+
+## Additional Resources
+
+[William Fiset Recursion Playlist](https://youtube.com/playlist?list=PLDV1Zeh2NRsCmu1lb9grUcljeYJtmgmYc&si=ZCzm6PNRa7iZw1bV)
```
Maaaybeeee take out the analytics ID in the YouTube link?

```suggestion
[William Fiset Recursion Playlist](https://youtube.com/playlist?list=PLDV1Zeh2NRsCmu1lb9grUcljeYJtmgmYc)
```

JacobGa03 replied:
Just another things I didn't realize, my bad. Once we decide on what links to include, then I will make some more changes.

jerrettl commented on src/cs1/recursion/recursion-basics.md:
```diff
@@ -0,0 +1,135 @@
+
+[Introduction to Recursion](https://www.geeksforgeeks.org/introduction-to-recursion-data-structure-and-algorithm-tutorials/)
+
+[5 Simple Steps for Solving Any Recursive Problem](https://www.youtube.com/watch?v=ngCos392W4w)
```
I really like this video's visual style! 3Blue1Brown's graphics library is killer.

JacobGa03 replied:
When linking things like videos, articles, etc., should we just put the title of the media we are linking? Or should we add some additional info like "Creator X's video "(insert video title)".

jerrettl commented on src/cs1/recursion/recursion-basics-answers.md:
```diff
@@ -0,0 +1,70 @@
+Here `n % 10` takes n and completes the modulus operation. Modulus, `%`, calculates the remainder when one number, `p`, is divided by another number, `q`. Here `n % 10` essentially pulls out the digit in the 1's place. Once we have pulled the digit out of the 1's place, we need to sum the rest of the digits in n. `sumDigits(n/10)` is our recursive call and calculates the sum of the rest of the di ...
+</details>
+
+## Lucas Numbers
```
This is fun! I like this example since it isn't just the standard Fibonacci example we see a lot.

JacobGa03 replied:
I thought it was really neat. I believe I got it from a Guha Lab or the FE, but I'm not sure where.

jerrettl reviewed (commented):
Hey Jacob!! Really nice work with all of this. I think these are pretty comprehensive as far as things go. Since most of the content itself I think is pretty clear, most of the comments I have will be logistics.

(I should note that I tend to be pretty picky, but only because I look at this as a full "product" that we are building. I aim to set students up for the best advantage they can get, which usually just comes down to consistency. I also tend to check grammar and spelling more than your usual code submission since the focus is on writing and explaining things. Any comments I make are purely on the work itself, and nothing on you as an author. It is a lot but I'm always happy to hear your thoughts!)

Also since there are changes in multiple PRs, the comments related to contribution standards will be mostly here, but if anything comes up on the others, I will edit the list here. (The other PR comments will be MUCH shorter.)

I have some overall thoughts and some specific things to take a look at first, then the rest is all stuff that will need to be ironed out for the sake of creating a contribution standard. (So this will be a bit longer as the guinea pig!) Linting will help with this a ton as well.

There will also be some other small comments for very very small things, most of which are subjective.

Comments in no specific order:

- Again, really nice work! I think these explanations are pretty well-described. I really like that you went through this by "building up" the factorial example, which shows *why* we have certain structures and patterns in place for recursion. There is really not much substantial I need to say on your PR overall.
- I notice that sometimes "Factorial" is used with the capital, and sometimes lowercase. Factorial is not a proper noun, so I wanted to check if there was intent for using this? Otherwise lowercase would be closer to how English prose is written.
- The structure of headings might just want to be looked at for the "Recursion Basics" page, since right now it looks like everything is inside the introduction section. So maybe for example the conclusion, or additional resources would be a level 1 heading. I do see you already do this in other PRs, so more of just a consistency thing.
- I have still been struggling with thinking about the best way to do this, but with the links included in the pages, I am weary of the "[evergreen content](https://github.com/wikiknights/cs-materials/wiki/Document-Contents#evergreen-content)" guideline. To be clear, I love adding the extra links to other resources!! This is really good! I have always also been cautious of adding links to outside pages due to the potential maintenance factor that comes with it if a link ever has to change.  I did make an [extra credits page](https://github.com/wikiknights/cs-materials/wiki/Extra-Credits) to start getting on this, but I'm open to thoughts on how to approach this. This would be part of contributor guidelines as well.
	* That said, I completely want to make sure all the credit and extra resources are there as needed. Maybe you can give the credits to Dr. Guha, but refer to the link outside the main content page? We can add the extra credits page to the site footer to keep it separate from any page too.
	* While I'm not sure how much that is the case now with the greater use of `<details>` tags these days, but the idea is that this could also export to PDF and be printed for personal notes. Of course you can't click a link on a piece of paper, but you can just omit the footer at the bottom on PDF export (which is what is done deep down in our PDF export template).

Contributor guidelines to consider:

- How should titles be capitalized? This is another thing that can be handled with a linter. Some examples of these standards are with the [Associated Press](https://www.codot.gov/business/grants/safetygrants/assets/APStyleGuideCheatSheet.pdf) or [APA](https://apastyle.apa.org/style-grammar-guidelines/capitalization/title-case).
- Some grammar and spelling rules should apply, so for example using the full (non-abbreviated) forms of terms where possible for clarity (unless it has already been mentioned previously on the page) and ending all sentences with punctuation.
- The names of the entries within the main table of contents (like the CS1 page) should match the titles of each respective page. It is a small thing, and you can choose either way, but this should be consistent, whichever title you choose.
- Space vs. tab indentation: The Markdown format specifies to use spaces for indentation. (There is no issue with this here, but writing it for reference.)
- In code blocks, should there be a space between the tick marks and the language name? This should be based on whatever the linter says. At least in this PR's case, there is some inconsistency, but I will consider this part of the linting discussion.
- How should we do numbering? Numbering in Markdown has typically been done with "1.", "2."... and so on, and "#." as special for magic numbering.
- We need a standard for code formatting specifically.
	* I'm mostly leaning towards the [CS50 Style Guide for C](https://cs50.readthedocs.io/style/c/), as a lot of inspiration for this project came from them and I think they're the most sensible. That said, the [GNU Coding Standards](https://www.gnu.org/prep/standards/standards.html) or the [Linux kernel coding style](https://www.kernel.org/doc/html/latest/process/coding-style.html) are other possible contenders. Definitely open to thoughts!
	* This also includes code indentation standards. (2 spaces, 4 spaces, etc.)
- Markdown should have a newline before any heading. (Some interpreters will have a weird time with this otherwise. I have unfortunately learned this from experience...)
- The name of the organization is "Wiki Knights". (Even our leadership has gotten this wrong from time to time, but that is the de-facto way to write it.)
- We are Americans (YEEHAW 🦅🏈🏈🤠🇺🇸) so we should specify American English spelling. (This article uses American English spelling, and I didn't notice any misspellings using that as a base.)
- Along with the evergreen guidelines, I have also considered this being "course-agnostic," meaning the pages with actual explanations don't refer to any one specific course, like COP 3223 or Introduction to C Programming. While we do have that within our main page organization and folder structure, I think the pages themselves should be completely self-enclosed.
- Where should we place questions/answers in practice problem pages? Should they be combined alongside the questions using `<details>`, or should they be on a separate page? I see you have them merged in your practice problems page. Originally, we've had solutions exclusively in a separate page, since we used this for classes and had the questions provided but didn't give the solutions right away. But this is another old aspect of the project that was never really defined.
- C specific: Use of `scanf()`, `printf()`, `gets()`, etc. for input and output. This is one of those parts of C programming that is always going to differ depending on who you ask. Generally, we use a certain set of patterns at UCF for working with strings, so we can just specify that.

The rest are all specific to parts of the code outlined as follows:

JacobGa03 committed:
Update src/cs1/recursion/recursion-basics.md

Co-authored-by: Jerrett Longworth <jerrett7@gmail.com>

JacobGa03 committed:
Update src/cs1/recursion/recursion-basics.md

Co-authored-by: Jerrett Longworth <jerrett7@gmail.com>

JacobGa03 committed:
Update src/cs1/recursion/recursion-basics.md

Co-authored-by: Jerrett Longworth <jerrett7@gmail.com>

JacobGa03 committed:
Update src/cs1/recursion/recursion-basics.md

Co-authored-by: Jerrett Longworth <jerrett7@gmail.com>

JacobGa03 committed:
Update src/cs1/recursion/recursion-basics.md

Co-authored-by: Jerrett Longworth <jerrett7@gmail.com>

JacobGa03 commented:
For the markdown linting I am using [markdownlint](https://marketplace.visualstudio.com/items?itemName=DavidAnson.vscode-markdownlint) as a VSCode extension. I do believe there is a CLI one can install to do a similar sort of thing for the giga-chads using Vim/NVim etc...

JacobGa03 committed:
Update src/cs1/recursion/recursion-basics.md

Co-authored-by: Jerrett Longworth <jerrett7@gmail.com>

JacobGa03 committed:
Small Trailing Spaces/Small Consistency Problems

Fixed some small issues with trailing spaces.

Also made it so ALL code snippets end at the case of fact(1) instead of fact(0). This was more for just consistency between the recursive and iterative solutions.

JacobGa03 committed:
Heading problem/Indentation

Rectified some issues with the headings. This still likely needs to be worked out with Jerrett.

Also fixed the indentation with the code snippets provided.

JacobGa03 committed:
Small Inconsistency Issues Fixed

JacobGa03 committed:
Practice/Answer Files Made to Match Intro to C

I split up the practice and answer pages into separate pages so they matched the format of the intro to c section.

JacobGa03 committed:
Removed unnecessary text at bottom
